services/publicapi/waiter_test.go

```diff
@@ -3,105 +3,209 @@ package publicapi
+			assert.NoError(t, err)
+			require.NotNil(t, wo, "wait object (2) is not nil when it should")
+			c <- struct{}{}
 		}()
 
-		w.reportCompleted(receipt, 0, 0)
+		waiter.deleteByChannel(wc1) // as if it was returned error quickly
+		waiter.complete(key, &waiterObject{"hello"})
+		<-c
 		<-c
+
+		_, open := <-wc1.c
+		require.False(t, open, "channel 1 should be closed")
+		_, open = <-wc2.c
+		require.False(t, open, "channel 2 should be closed")
 	})
 }
 
-func TestWaitForTransaction_Timeout(t *testing.T) {
+func TestPublicApiWaiter_CompleteChanWhenOtherIsTimedOut(t *testing.T) {
 	test.WithContext(func(ctx context.Context) {
-		w := newTxWaiter(ctx)
-		txHash := make(primitives.Sha256, 32)
-		rand.Read(txHash)
+		key := "key"
+		waiter := newWaiter(ctx)
+		wc1 := waiter.add(key)
+		wc2 := waiter.add(key)
 
-		waitContext := w.createTxWaitCtx(txHash)
-		defer waitContext.cleanup()
+		c := make(chan struct{}, 2)
+		go func() {
+			wo, err := waiter.wait(wc1, 5*time.Millisecond)
+			assert.Error(t, err)
+			require.Nil(t, wo, "wait object (1) should be nil")
+			c <- struct{}{}
+		}()
 
-		_, err := waitContext.until(1 * time.Millisecond)
+		go func() {
+			wo, err := waiter.wait(wc2, 1*time.Second)
+			assert.NoError(t, err)
+			require.NotNil(t, wo, "wait object (2) is not nil when it should")
+			c <- struct{}{}
+		}()
 
-		require.EqualError(t, err, "timed out waiting for transaction result", "Timeout did not occur")
+		time.Sleep(15 * time.Millisecond)
+		waiter.complete(key, &waiterObject{"hello"})
+		<-c
+		<-c
+
+		_, open := <-wc1.c
+		require.False(t, open, "channel 1 should be closed")
+		_, open = <-wc2.c
+		require.False(t, open, "channel 2 should be closed")
 	})
 }
 
-func TestWaitForTransaction_GracefulShutdownFreesAllWaitingGoroutines(t *testing.T) {
-	var w *txWaiter
-	done := make(chan struct{})
-	test.WithContext(func(ctx context.Context) {
-		w = newTxWaiter(ctx)
+func TestPublicApiWaiter_WaitGracefulShutdown(t *testing.T) {
+	var waiter *waiter
+	c := make(chan struct{})
 
-		var waitTillCancelled = func() {
-			txHash := make(primitives.Sha256, 32)
-			rand.Read(txHash)
-			waitContext := w.createTxWaitCtx(txHash)
-			defer waitContext.cleanup()
+	test.WithContext(func(ctx context.Context) {
+		key := "key"
+		waiter = newWaiter(ctx)
+		wc1 := waiter.add(key)
+		wc2 := waiter.add(key)
 
+		var waitTillCancelled = func(wc *waiterChannel) {
 			startTime := time.Now()
-			_, err := waitContext.until(1 * time.Second)
+			wo, err := waiter.wait(wc, 1*time.Second)
 			assert.Error(t, err, "expected waiting to be aborted")
 			assert.WithinDuration(t, time.Now(), startTime, 100*time.Millisecond, "expected not to reach timeout")
-			done <- struct{}{}
+			require.Nil(t, wo, "wait object (1) should be nil")
+			c <- struct{}{}
 		}
-		go waitTillCancelled()
-		go waitTillCancelled()
+
+		go waitTillCancelled(wc1)
+		go waitTillCancelled(wc2)
 	})
-	<-done
-	<-done
+	<-c
+	<-c
 }
+
+// TODO  MUTEX TESTS ? and/or timing tests
```

Comment:
Please review your tests - I guess this wasn't TDD, but I see that delete by key is never tested (For instance.)